migrate_cpp/cpp_refactoring/rewriter.h

```diff
@@ -0,0 +1,205 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_MIGRATE_CPP_CPP_REFACTORING_REWRITER_H_
+#define CARBON_MIGRATE_CPP_CPP_REFACTORING_REWRITER_H_
+
+#include <optional>
+#include <string>
+#include <utility>
+#include <variant>
+#include <vector>
+
+#include "clang/AST/ASTConsumer.h"
+#include "clang/AST/ASTTypeTraits.h"
+#include "clang/AST/RecursiveASTVisitor.h"
+#include "clang/Frontend/CompilerInstance.h"
+#include "clang/Frontend/FrontendAction.h"
+#include "llvm/ADT/DenseMap.h"
+#include "migrate_cpp/cpp_refactoring/output_segment.h"
+
+namespace Carbon {
+
+// `OutputWriter` is ressponsible for traversing the tree of `OutputSegment`s
+// and writing the correct data to its member `output`.
+class OutputWriter {
+  struct Empty {
+    friend bool operator==(Empty, Empty) { return true; }
+  };
+  struct Tombstone {
+    friend bool operator==(Tombstone, Tombstone) { return true; }
+  };
+  using KeyType =
+      std::variant<clang::DynTypedNode, clang::TypeLoc, Empty, Tombstone>;
+
+  struct KeyInfo {
+    static bool isEqual(const KeyType& lhs, const KeyType& rhs) {
+      return lhs == rhs;
+    }
+    static unsigned getHashValue(const KeyType& x) {
+      return std::visit(
+          [](auto x) -> unsigned {
+            using type = std::decay_t<decltype(x)>;
+            if constexpr (std::is_same_v<type, clang::DynTypedNode>) {
+              return clang::DynTypedNode::DenseMapInfo::getHashValue(x);
+            } else if constexpr (std::is_same_v<type, clang::TypeLoc>) {
+              // TODO: Improve this.
+              return reinterpret_cast<uintptr_t>(x.getTypePtr());
+            } else {
+              return 0;
+            }
+          },
+          x);
+    }
+
+    static KeyType getEmptyKey() { return Empty{}; }
+    static KeyType getTombstoneKey() { return Tombstone{}; }
+  };
+
+ public:
+  using SegmentMapType =
+      llvm::DenseMap<KeyType, std::vector<OutputSegment>, KeyInfo>;
+
+  auto Write(clang::SourceLocation loc, const OutputSegment& segment) const
+      -> bool;
+
+  const SegmentMapType& map;
+
+  // Bounds represent the offsets into the primary file (multi-file refactorings
+  // are not yet supported) that should be output. While primarily this is a
+  // mechanism to make testing more robust, it can also be used to make local
+  // changes to sections of C++ code.
+  std::pair<size_t, size_t> bounds;
+
+  clang::SourceManager& source_manager;
+  std::string& output;
+};
+
+// `RewriteBuilder` is a recursive AST visitor. For each node, it computes and
+// stores a sequence of `OutputSegment`s describing how this node should be
+// replaced.
+class RewriteBuilder : public clang::RecursiveASTVisitor<RewriteBuilder> {
```

Comment:
They don't really, but I think RewriteBuilder is going to be long no matter what as it has to hold the code for all nodes. I think I'd prefer to keep OutputWriter in the same file for now since it defines the segment map type that everything uses, but I'll definitely keep this in mind moving forward.